igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerBuildMonitor.groovy

```diff
@@ -0,0 +1,335 @@
+            log.debug("[${master}:${pipeline}] currentlyBuilding: ${currentlyBuilding}" )
+            log.debug("[${master}:${pipeline}]   completedBuilds: ${completedBuilds}" )
+            cursor = cursor?:lastBuildStamp
+            Date lowerBound = new Date(cursor)
+            if (!igorProperties.spinnaker.build.processBuildsOlderThanLookBackWindow) {
+                completedBuilds = onlyInLookBackWindow(completedBuilds)
+            }
+            delta.add(new PipelineDelta(
+                    cursor: cursor,
+                    name: pipeline,
+                    lastBuildStamp: lastBuildStamp,
+                    upperBound: upperBound,
+                    lowerBound: lowerBound,
+                    completedBuilds: completedBuilds,
+                    runningBuilds: currentlyBuilding
+                    ))
+        } catch (e) {
+            log.error("Error processing runs for [{}:{}]", kv("master", master), kv("pipeline", pipeline), e)
+            if (e.cause instanceof RetrofitError) {
+                def re = (RetrofitError) e.cause
+                log.error("Error communicating with Wercker for [{}:{}]: {}", kv("master", master), kv("pipeline", pipeline), kv("url", re.url), re)
+            }
+        }
+    }
+
+    private List<Run> onlyInLookBackWindow(List<Run> builds) {
+        use(TimeCategory) {
+            def offsetSeconds = pollInterval.seconds
+            def lookBackWindowMins = igorProperties.spinnaker.build.lookBackWindowMins.minutes
+            Date lookBackDate = (offsetSeconds + lookBackWindowMins).ago
+            return builds.stream().filter({
+                Date buildEndDate = it.finishedAt
+                return buildEndDate.after(lookBackDate)
+            }).collect(Collectors.toList())
+        }
+    }
+
+    private GenericBuild toBuild(String master, String pipeline, Run run) {
+        Result res = (run.finishedAt == null) ? Result.BUILDING : (run.result.equals("passed")? Result.SUCCESS : Result.FAILURE)
+        return new GenericBuild (
+                building: (run.finishedAt == null),
+                result: res,
+                number: cache.getBuildNumber(master, pipeline, run.id),
+                timestamp: run.startedAt.fastTime as String,
+                id: run.id,
+                url: run.url
+                )
+    }
+
+    @Override
+    protected void commitDelta(PipelinePollingDelta delta, boolean sendEvents) {
+        String master = delta.master
+        delta.items.parallelStream().forEach { pipeline ->
+            //job = Wercker pipeline (org/app/pipeline)
+            // post event for latest finished run
+            pipeline.completedBuilds.forEach { run ->
+                //build = Wercker run
+                Boolean eventPosted = cache.getEventPosted(master, pipeline.name, run.id)
+                GenericBuild build = toBuild(master, pipeline.name, run)
+                if (!eventPosted && echoService.isPresent() && sendEvents) {
```

Comment:
Is there a reason to gate this on `echoService.isPresent()`; this is already checked in `postEvent` and throws an error if Echo isn't present (which seems desirable since this won't work without Echo).

Reply:
Fixed. Originally when echoService is NOT present, cache.setEventPosted may still get called; this may still be an issue in Jenkins as we used Jenkins as example.